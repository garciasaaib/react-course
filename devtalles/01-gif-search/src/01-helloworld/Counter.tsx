import { useState } from "react";

export const Counter = ({initialValue = 0}) => {
  const [counter, setCounter] = useState(initialValue);
  const handleAdd = () => {
    setCounter((currValue) => currValue+1);
  }
  const handleSubstract = () => {
    setCounter((currValue) => currValue-1);
  }
  const handleReset = () => {
    setCounter(0);
  }
  return (
    <div>
      <h2>Counter: {counter}</h2>
      <button onClick={handleAdd}>Add</button>
      <button onClick={handleReset}>Reset</button>
      <button onClick={handleSubstract}>Substract</button>

    </div>
  )
}
