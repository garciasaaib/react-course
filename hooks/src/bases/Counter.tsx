import { useState } from "react";

interface Props {
    initialValue?: number
}

export const Counter = ({ initialValue = 0 }: Props) => {

    const [counter, setCounter] = useState(initialValue)

    const handleClick = () => {
        setCounter( prev => prev + 1 );
    }

    return (
        <>
            <h3>Counter: { counter }</h3>   

            <button onClick={ handleClick }>
                +1
            </button>
        </>
    )
}
