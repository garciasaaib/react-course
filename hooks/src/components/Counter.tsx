import { useState } from "react";

export const Counter = () => {
	const [count, setCount] = useState(0);

	const incrementar = (num: number) => {
		setCount(count + num);
	}

	return (
		<>
			<h2>UseState counter is {count}</h2>
			<div className="card">
				<button onClick={() => incrementar(1)}>+1</button>
				<button onClick={() => incrementar(2)}>+2</button>
				<button onClick={() => setCount(0)}>0</button>

			</div>


		</>
	);
};
