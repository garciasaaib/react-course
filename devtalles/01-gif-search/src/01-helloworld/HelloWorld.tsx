import { Counter } from "./Counter";

// un componente padre
export const HelloWorld = ({ title = "holamundo" }) => {
	return (
		<section>
			<Title title={title} />
			<Counter initialValue={1} />
		</section>
	);
};

// un componente hijo
export const Title = ({ title = "hola soy goku" }: { title?: string }) => (
	<h1
		style={{ background: "teal" }}
		className="custom inline"
		data-testid="test-title"
	>
		{title}
	</h1>
);

/** Notas
 * - <></> es un fragmento de React, es como un div pero no se renderiza en el DOM
 * - Se puede usar React.Fragment o <>
 * - Las variables y funciones estaticas pueden vivir preferentemente fuera del componente
 * - Los styles se agregan como un objeto style al componente se agregan con className o con un objeto
 * - props son un objeto que permiten la comunicacion entre componente padre e hijo
 * - DefaultProps es una propiedad que permite definir valores por defecto a las props, eso las hace opcionales
 * - con proptypes puedes definir el tipo de dato que esperas recibir, pero si tienes ts no necesitas proptypes
 */
