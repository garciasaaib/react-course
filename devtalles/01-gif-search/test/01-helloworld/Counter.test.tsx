import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Counter } from "../../src/01-helloworld/Counter";

describe('<Counter />', () => {
  test('Debe mostrar <Counter /> correctamente', () => {
		const { container } = render(<Counter />);
		expect(container).toMatchSnapshot();
  });

  test('Debe mostrar el valor por defecto de 100', () => {
    render(<Counter initialValue={100} />);
    // obteneos los tag h2, y evaluamos su valor
    expect(screen.getByRole('heading', {level: 2}).innerHTML).toContain('100');
  });


  test('Debe 101 despues del click en Add', () => {
    const initialValue = 100;
    render(<Counter initialValue={initialValue} />);
    // obtenemos el elemento button con el valor Add, y le hacemos un click
    fireEvent.click(screen.getByRole('button', {name: 'Add'}));
    // obtenemos el h2 y evaluamos su contenido
    expect(screen.getByRole('heading', {level: 2}).innerHTML).toContain('101');
  });

  test('Debe 99 despues del click en Add', () => {
    const initialValue = 100;
    render(<Counter initialValue={initialValue} />);
    // obtenemos el boton con el texto Add, y hacemos click
    fireEvent.click(screen.getByRole('button', {name: 'Add'}));
    // evaluamos el contenido del h2
    expect(screen.getByRole('heading', {level: 2}).innerHTML).toContain('101');
  });
});