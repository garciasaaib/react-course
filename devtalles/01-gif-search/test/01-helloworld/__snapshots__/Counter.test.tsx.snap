// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> Debe mostrar <Counter /> correctamente 1`] = `
<div>
  <div>
    <h2>
      Counter: 
      0
    </h2>
    <button>
      Add
    </button>
    <button>
      Reset
    </button>
    <button>
      Substract
    </button>
  </div>
</div>
`;
